server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -523,7 +523,13 @@ private <T> void ensureAfterSeqNoRefreshed(
                     );
                 }
             }, timeout, Names.SAME);
-            shard.addRefreshListener(waitForCheckpoint, new ActionListener<>() {
+
+            // allow waiting for not-yet-issued sequence number if shard isn't promotable to primary and the timeout is less than or equal
+            // to 30s
+            final boolean allowWaitForNotYetIssued = shard.routingEntry().isPromotableToPrimary() == false
```

Comment:
We have to work this out here and pass a flag down the stack because it's only here that we know about the timeout.